src/broadcast.jl

```diff
@@ -110,7 +111,7 @@ scalar_getindex(x::Tuple{<: Any}) = x[1]
     exprs = similar(indices, Expr)
     for (j, current_ind) ∈ enumerate(indices)
         exprs_vals = [
-            (!(a[i] <: AbstractArray) ? :(scalar_getindex(a[$i])) : :(a[$i][$(broadcasted_index(sizes[i], current_ind))]))
+            (!(a[i] <: AbstractArray || a[i] <: Tuple) ? :(scalar_getindex(a[$i])) : :(a[$i][$(broadcasted_index(sizes[i], current_ind))]))
```

Comment:
I guess this means we should kill off the `scalar_getindex(x::Tuple)` further up? TBH I'm not sure why that was ever there.

Reply:
We could, it looks like it was there to allow broadcasting with one-element tuples.

Reply:
Ok, I think we should remove it for consistency, and add some tests to check that the functionality still works. AFAICT cases like

```julia
SA[1,2,3] .* (SA[1,0,0],)
```

used to work because of the `scalar_getindex` hack and your changes preserves that functionality while also allowing more things to work.

TBH the fact that we check `a[i] <: AbstractArray` makes me nervous and it seems like these problems will recur with other user-defined types (because we may not respect the `broadcastable` API? I'm not an expert on how this is set up though).

Regardless of that what you've got here is a definite improvement so we should go with it.

Reply:
> TBH the fact that we check `a[i] <: AbstractArray` makes me nervous and it seems like these problems will recur with other user-defined types (because we may not respect the `broadcastable` API? I'm not an expert on how this is set up though).

Yes, it's a bad hack that will bite us in the future but as far as I understand the broadcast machinery (I'm not an expert though) this unrolled broadcasting can't be expected to work with arbitrary types participating in broadcasting anyway. A potentially better solution would be checking `Base.BroadcastStyle` of `a[i]` for `Style{Tuple}` and `AbstractArrayStyle`, so that could be done as well. But constructing `Base.BroadcastStyle` in a generated function doesn't sound like a great idea either. Or we could construct `Base.BroadcastStyle` in `Base.copy(B::Broadcasted{StaticArrayStyle{M}})`.

This can be improved gradually so we can make this a little bit better in this PR first.